### PR TITLE
[Feature-3934] Add registy module for streampark

### LINKED ISSUE #3934: [Feature] Add registy module for streampark
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-streampark/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

Registrar SPI, the user can specify the registration center method and specific related parameters in the configuration file. Zookeeper is used by default

todo list:

1. Provide a unified `RegistryClient` related interface (org.apache.streampark.registry), Including registration, offline, update and other operations. The external access registration center uniformly calls the Client interface.
The registration center currently only provides two registration center(MySql & Zookeeper). We need to load the corresponding plug-in according to the user configuration. When the user does not specify it, zookeeper is used by default.

At the same time, in order to transition, we need to provide the implementation of distributed locks. (This is just for transition)

2. Implementation of Zookeeper in the registry

3. Implementation of MySql in the registry 

4. Console Server supports the registration center. The console server must share the same registry with each other, and multiple registries are not currently supported.

## Function

SP use registry to do the below three things:

1. Store the metadata of console so that it can get notify when nodes up and down.
2. Store the metadata of console to do load balance.
3. Acquire a global lock when do failover.

So for SP, the registry need to notify the server when the server subscribe data have added/deleted/updated, support a way to create/release a global lock,
delete the server's metadata when server down.

## How to use

At present, we have implements two registry: Zookeeper(Default),Mysql. If you
want to use them, you should config it at resource/application.yaml. The configuration details
can be viewed in the README of plugin under Module streampark-registry-plugins

### streampark-registry-all

This module is used for exporting the implementation of registry.
If you want to add new registry,you should add the dependency in the pom.xml

### streampark-registry-api

This module contains the relevant interfaces involved in the use of the registry.
The following are several important interfaces
1. Registry Interface: If you want to implement your own registry, you just need to implement this interface
2. ConnectionListener Interface: This interface is responsible for the connection status between the client and the registry,
The connection state can be viewed in ConnectionState.java
3. SubscribeListener Interface: This interface is responsible for monitoring the state changes of child nodes under the specified prefix.
Event content can be viewed in event.java

### streampark-registry-plugins

This module contains all registry implementations in SP

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by HxpSerein**
## What changes were proposed in this pull request

Issue Number: close https://github.com/apache/incubator-streampark/issues/3934

## Brief change log
Add streampark-registy module

## Verifying this change

This change added tests and can be verified as follows:
- NetworkUtilsTestCase
- UUIDUtilsTestCase
- RegistryTestCase
- JdbcRegistryTestCase
- JdbcRegistryThreadFactoryTest
- LockUtilsTest
- MysqlJdbcRegistryTestCase
- PostgresqlJdbcRegistryTestCase

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): yes

### CODE REVIEW

File: streampark-common/src/main/java/org/apache/streampark/common/enums/ServerStatus.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.streampark.common.enums;
+
+public enum ServerStatus {
```
Comment (wolfboys) [inline streampark-common/src/main/java/org/apache/streampark/common/enums/ServerStatus.java]: streampark-common is a global module. Avoid putting console layer logic into streampark-common.

File: streampark-common/src/main/java/org/apache/streampark/common/lifecycle/ServerLifeCycleException.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.streampark.common.lifecycle;
+
+public class ServerLifeCycleException extends Exception {
```
Comment (wolfboys) [inline streampark-common/src/main/java/org/apache/streampark/common/lifecycle/ServerLifeCycleException.java]: This logic should not be placed in the common module; it belongs in the console-service layer.

File: streampark-common/src/main/java/org/apache/streampark/common/enums/ServerStatus.java (hunk 1 above)
Comment (HxpSerein) [reply streampark-common/src/main/java/org/apache/streampark/common/enums/ServerStatus.java]: I have moved the relevant classes from `streampark-common` to `streampark-console-registry`.

File: streampark-common/src/main/java/org/apache/streampark/common/lifecycle/ServerLifeCycleException.java (hunk 2 above)
Comment (HxpSerein) [reply streampark-common/src/main/java/org/apache/streampark/common/lifecycle/ServerLifeCycleException.java]: Moved.

**Review by wolfboys — APPROVED** [review]
LGTM

**Comment by ruanwenjun** [comment]
@wolfboys @HxpSerein Hi, I find most of the code under `streampark-console-registry` is copy from DS, you should announce this.

**Comment by HxpSerein** [comment]
@ruanwenjun Thank you for reporting this issue. After review, we confirmed that the issue does exist. This code is only in the dev branch. We will list the code in the LICENSE. This is only a temporary solution and we will rewrite this part of the code later. Thank you again for your feedback.

**Comment by wolfboys** [comment]
hi @ruanwenjun @HxpSerein @SbloodyS:

This pull request is a fork of [another PR](https://github.com/apache/incubator-streampark/pull/3948), After review, we discovered that the code in the `streampark-console-registry` module was copied from [Apache DolphinScheduler](https://github.com/apache/dolphinscheduler/tree/dev/dolphinscheduler-registry) and not announce this, This violates the Apache License specifications. Therefore, we have submitted a [new PR](https://github.com/apache/incubator-streampark/pull/4122) to reverse the changes made in the previous request. We are very sorry for any inconvenience this may cause.

**Pull request merged**